genie-web/src/main/java/com/netflix/genie/web/util/StreamBuffer.java

```diff
@@ -0,0 +1,199 @@
+/*
+ *
+ *  Copyright 2019 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.web.util;
+
+import com.google.protobuf.ByteString;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.NotImplementedException;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A temporary in-memory structure to hold in-transit data.
+ * Provides an {@code InputStream} for reading, reading blocks until data becomes available or the buffer is closed.
+ * <p>
+ * To avoid in-memory data growing excessively, this buffer stores a single "chunk" at the time.
+ * Only after a chunk is consumed, a new one can be appended.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@ThreadSafe
+@Slf4j
+public class StreamBuffer {
```

Comment:
Seems similar to what we're doing with DAS. May be a good use case for a Reactive pipeline but that's more of a "if we feel like looking into that". This works perfectly fine